### PR TITLE
Add gradient customization

### PR DESCRIPTION
**Pull request by rasmustautsglia**
Currently there was no way to apply gradient colors to the Theme either directly or through remote configuration. With this PR, gradient can be applied to view and button backgrounds. Border, and foreground remain with 1 color.

MOB-1588

### CODE REVIEW

**Comment by rasmustautsglia** [comment]
Simulator Screen Shot - iPhone 13 Pro Max - 2022-10-12 at 17 35 44
Simulator Screen Shot - iPhone 13 Pro Max - 2022-10-12 at 11 47 54
Simulator Screen Shot - iPhone 13 Pro Max - 2022-10-12 at 11 47 37

**Comment by rasmustautsglia** [comment]
Survey does not render correctly

File: GliaWidgets/ColorType.swift [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import UIKit
+
+public enum ColorType {
+    case fill(color: UIColor)
+    case gradient(colors: [CGColor])
+
+    static func == (lhs: ColorType, rhs: ColorType) -> Bool {
+        switch (lhs, rhs) {
```
Comment (dukhovnyi) [inline GliaWidgets/ColorType.swift:8]: I'd like to clarify, why you didn't use `default`, do you think that we may extend this enum in future?

File: GliaWidgets/Lib/Extensions/UIView+Extensions.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import UIKit
+
+private let reusableIdentifier = "theme.reusableIdentifier.background"
+
+extension UIView {
+    func makeGradientBackground(colors: [CGColor], cornerRadius: CGFloat?) -> CAGradientLayer {
+        let l: CAGradientLayer
+
```
Comment (igorkravchenko) [inline GliaWidgets/Lib/Extensions/UIView+Extensions.swift:8]: I would add comments explaining on what is going on here.

File: GliaWidgets/ViewController/Survey/SurveyViewController.View.swift [hunk 3]
```diff
@@ -103,42 +103,59 @@ extension Survey {
             ])
         }
 
+        override func layoutSubviews() {
+            super.layoutSubviews()
+
+            _updateUi?()
```
Comment (igorkravchenko) [inline GliaWidgets/ViewController/Survey/SurveyViewController.View.swift:109]: Let's add explanation about this. Also what is the reason for `_` ?

**Review by igorkravchenko — COMMENTED** [review]
Left a few comments. Let's also make sure that snapshot tests are up-to-date with theses changes.

File: GliaWidgets/ViewController/Survey/Components/ButtonView/Survey.ButtonView.swift [hunk 4]
```diff
@@ -62,10 +62,10 @@ extension Survey {
         override func defineLayout() {
             super.defineLayout()
             NSLayoutConstraint.activate([
-                value.topAnchor.constraint(equalTo: topAnchor, constant: 1),
-                value.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 1),
-                value.trailingAnchor.constraint(equalTo: trailingAnchor, constant: 1),
-                value.bottomAnchor.constraint(equalTo: bottomAnchor, constant: 1)
+                value.topAnchor.constraint(equalTo: topAnchor, constant: 0),
```
Comment (EgorovEI) [inline GliaWidgets/ViewController/Survey/Components/ButtonView/Survey.ButtonView.swift:65]: Do we need to remove 1pt paddings?

File: GliaWidgets/ViewController/Survey/Components/InputQuestionView/Survey.InputQuestionView.swift [hunk 5]
```diff
@@ -70,13 +70,18 @@ extension Survey {
...
-                UIColor(hex: style.option.normalLayer.borderColor).cgColor
+                style.option.highlightedLayer.borderColor :
+                style.option.normalLayer.borderColor
             textView.layer.borderWidth = props.showValidationError ?
                 style.option.highlightedLayer.borderWidth :
                 style.option.normalLayer.borderWidth
             if let backgroundColor = style.background.background {
-                textView.backgroundColor = UIColor(hex: backgroundColor)
+                switch backgroundColor {
+                case .fill(color: let color):
```
Comment (EgorovEI) [inline GliaWidgets/ViewController/Survey/Components/InputQuestionView/Survey.InputQuestionView.swift:80]:
```suggestion
                case .fill(let color):
```

File: GliaWidgets/ViewController/Survey/Components/InputQuestionView/Survey.InputQuestionView.swift [hunk 6]
```diff
@@ -70,13 +70,18 @@ extension Survey {
...
+                    textView.backgroundColor = color
+                case .gradient(colors: let colors):
```
Comment (EgorovEI) [inline GliaWidgets/ViewController/Survey/Components/InputQuestionView/Survey.InputQuestionView.swift:82]:
```suggestion
                case .gradient(let colors):
```

File: GliaWidgets/ViewController/Survey/Components/ButtonView/Survey.ButtonView.swift (hunk 4 above)
Comment (rasmustautsglia) [reply GliaWidgets/ViewController/Survey/Components/ButtonView/Survey.ButtonView.swift:65]: Yes, it made the offset visible below border

**Comment by EgorovEI** [comment]
Looks good! But [there](https://github.com/salemove/ios-sdk-widgets/pull/396/files#r1009259010) is one small comment from me

File: GliaWidgets/Lib/Extensions/UIView+Extensions.swift [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        if let index = layer.sublayers?.firstIndex(where: { $0.name == reusableIdentifier }),
+           let existedLayer = layer.sublayers?[index] as? CAGradientLayer {
+            l = existedLayer
+        } else {
+            l = CAGradientLayer()
+        }
+        l.name = reusableIdentifier
+        l.frame = self.bounds
```
[the remark this answers is not in the document]
Comment (rasmustautsglia) [reply GliaWidgets/Lib/Extensions/UIView+Extensions.swift:18]: fixed [1d02344](https://github.com/salemove/ios-sdk-widgets/pull/396/commits/1d02344b52eaa6280270323a07535b5dd790ed7b)

**Review by EgorovEI — APPROVED** [review]

**Review by igorkravchenko — APPROVED** [review]
I approve it since I don't have a better solution adding gradient support for Survey. However I think it is worth to review current implementation during cooldown.

**Comment by rasmustautsglia** [comment]
!merge

**Pull request merged**